qadence/ml_tools/callbacks/callbackmanager.py

```diff
@@ -0,0 +1,216 @@
+            self.add_callback("LogHyperparameters", "on_train_end")
+        # Log model
+        if self.config.log_model:
+            self.add_callback("LogModelTracker", "on_train_end")
+        # only save the last checkpoint if not checkpoint_best_only
+        if not self.config.checkpoint_best_only:
+            self.add_callback("SaveCheckpoint", "on_train_end")
+        self.add_callback("WriteMetrics", "on_train_end")
+
+    def add_callback(self, callback: Union[Callback, str], on: str, called_every: int = 1) -> None:
+        """
+        Adds a callback to the manager.
+
+        Args:
+            callback (Union[Callback, str]): The callback instance or name.
+            on (str): The event on which to trigger the callback.
+            called_every (int): Frequency of callback calls in terms of iterations.
+        """
+        if isinstance(callback, str):
+            callback_class = self.get_callback_by_name(callback)
+            if callback_class:
+                # Create an instance of the callback class
+                callback_instance = callback_class(on=on, called_every=called_every)
+                self.callbacks.append(callback_instance)
+            else:
+                logger.warning(f"Callback '{callback}' not recognized and will be skipped.")
+        elif isinstance(callback, Callback):
+            callback.on = on
+            callback.called_every = called_every
+            self.callbacks.append(callback)
+        else:
+            logger.warning(
+                f"Invalid callback type: {type(callback)}. Expected str or Callback instance."
+            )
+
+    def get_callback_by_name(self, name: str) -> Optional[Callable]:
+        """
+        Retrieves a callback class by name.
+
+        Args:
+            name (str): The name of the callback.
+
+        Returns:
+            Optional[Callable]: The callback class if found, else None.
+        """
+        callback_map = {
+            "PrintMetrics": PrintMetrics,
+            "WriteMetrics": WriteMetrics,
+            "PlotMetrics": PlotMetrics,
+            "SaveCheckpoint": SaveCheckpoint,
+            "LoadCheckpoint": LoadCheckpoint,
+            "LogModelTracker": LogModelTracker,
+            "LogHyperparameters": LogHyperparameters,
+            "SaveBestCheckpoint": SaveBestCheckpoint,
+        }
+        return callback_map.get(name)
+
+    def run_callbacks(self, trainer: Any) -> Any:
+        """
+        Runs callbacks that match the current training state.
+
+        Args:
+            trainer (Any): The training object managing the training process.
+
+        Returns:
+            Any: Results of the executed callbacks.
+        """
+        return [
+            callback(when=trainer.state, trainer=trainer, config=self.config, writer=self.writer)
+            for callback in self.callbacks
+            if callback.on == trainer.state
+        ]
+
+    def start_training(self, trainer: Any) -> None:
```

Comment:
Why Any?

Reply:
This is because of circular imports between Trainer and CallbackManager.